src/vcpkg/base/downloads.cpp

```diff
@@ -1152,144 +531,102 @@ namespace vcpkg
+        curl_easy_setopt(curl, CURLOPT_XFERINFOFUNCTION, &progress_callback);
+        curl_easy_setopt(curl, CURLOPT_XFERINFODATA, &machine_readable_progress);
+
+        // Retry on transient errors:
+        // Transient error means either: a timeout, an FTP 4xx response code or an HTTP 408, 429, 500, 502, 503 or 504
+        // response code.
+        bool curl_success = false;
+        bool should_retry = true;
+        size_t retries_count = 0;
+
+        using namespace std::chrono_literals;
+        static constexpr std::array<std::chrono::seconds, 3> retry_delay = {0s, 1s, 2s};
+        do
+        {
+            // blocking transfer
+            should_retry = false;
+            std::this_thread::sleep_for(retry_delay[retries_count++]);
+            auto curl_code = curl_easy_perform(curl);
+            if (curl_code == CURLE_OK)
+            {
+                long response_code = -1;
+                if (CURLE_OK == curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code))
                 {
-                    if (!download_winhttp(context,
-                                          machine_readable_progress,
-                                          fs,
-                                          download_path_part_path,
-                                          *split_uri_view,
-                                          sanitized_url))
+                    if ((response_code >= 200 && response_code < 300) || response_code == 0)
                     {
-                        return DownloadPrognosis::NetworkErrorProxyMightHelp;
+                        curl_success = true;
+                        break;
                     }
-
-                    if (!check_downloaded_file_hash(
-                            context, fs, sanitized_url, download_path_part_path, maybe_sha512, out_sha512))
+                    else if (response_code == 429 || response_code == 408 || response_code == 500 ||
+                             response_code == 502 || response_code == 503 || response_code == 504)
                     {
-                        return DownloadPrognosis::OtherError;
+                        should_retry = true;
+                        context.report_error(msg::format(msgCurlFailedHttpResponseWithRetry,
+                                                         msg::exit_code = static_cast<int>(curl_code),
+                                                         msg::count = retries_count,
+                                                         msg::value = retry_delay.size()));
+                    }
+                    else
+                    {
+                        context.report_error(
+                            msg::format(msgCurlFailedHttpResponse, msg::exit_code = static_cast<int>(curl_code)));
                     }
-
-                    fs.rename(download_path_part_path, download_path, VCPKG_LINE_INFO);
-                    return DownloadPrognosis::Success;
                 }
```

Comment:
I think you're missing an else branch here. (If curl_easy_perform succeeds but curl_easy_getinfo fails there probably needs to be an error report?)

Either that or go to Checks::unreachble in which case you can invert the condition and save an indent